data/model/namespacequota.py

```diff
@@ -0,0 +1,380 @@
+import json
+
+from peewee import fn, JOIN
+
+from data.database import (
+    Repository,
+    QuotaLimits,
+    UserOrganizationQuota,
+    QuotaType,
+    Manifest,
+    get_epoch_timestamp_ms,
+    Tag,
+    RepositorySize,
+)
+from data.model import (
+    InvalidOrganizationException,
+    _basequery,
+    organization,
+    user,
+    InvalidUsernameException,
+    repository,
+    notification,
+)
+
+
+def verify_namespace_quota(namespace_name, repository_ref):
+    repository.get_repository_size_and_cache(repository_ref._db_id)
+    namespace_size = get_namespace_size(namespace_name)
+    return check_limits(namespace_name, namespace_size)
+
+
+def verify_namespace_quota_force_cache(namespace_name, repository_ref):
+    force_cache_repo_size(repository_ref)
+    namespace_size = get_namespace_size(namespace_name)
+    return check_limits(namespace_name, namespace_size)
+
+
+def verify_namespace_quota_during_upload(namespace_name, repository_ref):
+    size = __gather_size_by_chunk_upload(repository_ref._db_id)
+    namespace_size = get_namespace_size(namespace_name)
+    if namespace_size is None:
+        namespace_size = 0
+
+    return check_limits(namespace_name, size + namespace_size)
+
+
+def check_limits(namespace_name, size):
+    limits = __get_namespace_limits(namespace_name)
+    limit_bytes = 0
+    severity_level = None
+    for limit in limits:
+        if size > limit["bytes_allowed"]:
+            if limit_bytes < limit["bytes_allowed"]:
+                limit_bytes = limit["bytes_allowed"]
+                severity_level = limit["type_id"]
+
+    return {"limit_bytes": limit_bytes, "severity_level": severity_level}
+
+
+def __gather_size_by_chunk_upload(repo_id):
+    return repository.get_size_during_upload(repo_id)
```

Comment:
nit: I initially thought this op was on a repository reference. Maybe `from data import model` instead, so that the reference is `model.repository.[..]` to make it clearer

Reply:
easy enough actually.